File: g/etl/store.go
package etl

import (
	"context"
	"database/sql"
)

type Store struct {
	db *sql.DB
}

func NewStore(db *sql.DB) *Store {
	return &Store{
		db: db,
	}
}

func (s *Store) InsertBatch(ctx context.Context, records []TransformedRecord) error {
	tx, err := s.db.BeginTx(ctx, nil)
	if err != nil {
		return err
	}
	defer tx.Rollback()

	stmt, err := tx.PrepareContext(ctx, `
		INSERT INTO processed_records (id, fullname, email_domain, source, ref_id, balance, processed_at)
		VALUES ($1, $2, $3, $4, $5, $6, $7)
	`)
	if err != nil {
		return err
	}
	defer stmt.Close()

	for _, r := range records {
		_, err := stmt.ExecContext(ctx,
			r.ID,
			r.FullName,
			r.EmailDomain,
			r.Source,
			r.RefID,
			r.Balance,
			r.ProcessedAt,
		)
		if err != nil {
			return err
		}
	}

	return tx.Commit()
}

File: z/src/etl/store.zig
const std = @import("std");
const Etl = @import("Etl.zig");
const pg = @import("pg");

pub fn insertBatch(pg_pool: *pg.Pool, records: std.ArrayList(Etl.TransformedRecord)) !void {
    const conn = try pg_pool.acquire();
    defer pg_pool.release(conn);

    _ = try conn.exec("BEGIN", .{});

    errdefer _ = conn.exec("ROLLBACK", .{}) catch |err| {
        std.log.err("rollback failure: {any}", .{err});
    };

    for (records.items) |r| {
        _ = try conn.exec(
            \\INSERT INTO processed_records (id, fullname, email_domain, source, ref_id, balance, processed_at)
            \\VALUES ($1, $2, $3, $4, $5, $6, $7)
        , .{
            r.id,
            r.full_name,
            r.email_domain,
            r.source,
            r.ref_id,
            r.balance,
            r.processed_at,
        });
    }

    _ = try conn.exec("COMMIT", .{});
}
